lib/Dialect/AIEVec/Transforms/VectorToAIEVecConversions.cpp

```diff
@@ -1207,24 +1348,40 @@ struct LowerVectorReductionOp
     if (kind == vector::CombiningKind::MINSI ||
         kind == vector::CombiningKind::MINUI ||
         kind == vector::CombiningKind::MINF) {
-      generateAIEVecOpsForReductionOp<aievec::MinOp>(rewriter, srcOp,
-                                                     shiftIndex);
+      generateAIEVecOpsForReductionOp<aievec::MinOp>(
+          rewriter, srcOp, shiftIndex, srcOp.getVector());
       return success();
     }
 
     // Reduction maximum
     if (kind == vector::CombiningKind::MAXSI ||
         kind == vector::CombiningKind::MAXUI ||
         kind == vector::CombiningKind::MAXF) {
-      generateAIEVecOpsForReductionOp<aievec::MaxOp>(rewriter, srcOp,
-                                                     shiftIndex);
+      generateAIEVecOpsForReductionOp<aievec::MaxOp>(
+          rewriter, srcOp, shiftIndex, srcOp.getVector());
       return success();
     }
 
     // Reduction add for i32, i16 and i8 types
     if (kind == vector::CombiningKind::ADD && isa<IntegerType>(scalarType)) {
-      generateAIEVecOpsForReductionOp<aievec::AddElemOp>(rewriter, srcOp,
-                                                         shiftIndex);
+      if (laneSize == 32 && elWidth == 32) {
+        Location loc = srcOp.getLoc();
+        VectorType vecType = createVectorType(laneSize / 2, scalarType);
+
+        auto lExtOp =
+            rewriter.create<aievec::ExtOp>(loc, vecType, srcOp.getVector(), 0);
+        auto rExtOp =
+            rewriter.create<aievec::ExtOp>(loc, vecType, srcOp.getVector(), 1);
+        auto addElemOp = rewriter.create<aievec::AddElemOp>(
+            loc, lExtOp.getResult().getType(), lExtOp.getResult(),
+            rExtOp.getResult());
+        shiftIndex /= 2;
+        generateAIEVecOpsForReductionOp<aievec::AddElemOp>(
+            rewriter, srcOp, shiftIndex, addElemOp.getResult());
+        return success();
+      }
+      generateAIEVecOpsForReductionOp<aievec::AddElemOp>(
+          rewriter, srcOp, shiftIndex, srcOp.getVector());
```

Comment:
All of this pattern is actually a bunch of different patterns that act on different operand types for the same op. Instead of having a long list of `if` statements within the same pattern (with a lot of code that will never execute because of its templated parameters), it's better to have one pattern per conversion. Each "replaceOpWithNewOp" or "generateAIEVecOpsForReductionOp" should live in its own individual pattern.